server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -690,6 +691,7 @@ public void onFailure(Exception e) {
             shardStateUpdated.countDown();
         }
         if (currentRouting.active() == false && newRouting.active()) {
+            checkAndCallWaitForStartedOrClosedShardListeners();
```

Comment:
Could we do this in `innerOpenEngineAndTranslog`?

This slot is a bit late, since we could respond immediately if the term/gen that we recover from is ok. Also, doing this here means it could run the listener on the cluster applier thread (not sure if that is problem).